app/lib/features/pins/pin_utils/pin_utils.dart

```diff
@@ -0,0 +1,174 @@
+import 'dart:io';
+import 'dart:typed_data';
+import 'package:acter/common/utils/utils.dart';
+import 'package:acter/common/widgets/attachment_options.dart';
+import 'package:acter/features/pins/providers/pins_provider.dart';
+import 'package:acter_flutter_sdk/acter_flutter_sdk_ffi.dart';
+import 'package:file_picker/file_picker.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:image_picker/image_picker.dart';
+import 'package:mime/mime.dart';
+
+typedef PinAttachment = ({AttachmentType type, File file});
+
+class PinUtils {
+  // belongs separately, useful for showing file picker directly (desktop)
+  static void _onTapFileSelection(WidgetRef ref) async {
+    FilePickerResult? result = await FilePicker.platform.pickFiles(
+      type: FileType.any,
+      allowMultiple: true,
+    );
+    if (result != null) {
+      List<PinAttachment> newAttachments =
+          result.paths.map<PinAttachment>((path) {
+        var file = File(path!);
+        var pinAttachment = (type: AttachmentType.file, file: file);
+        return pinAttachment;
+      }).toList();
+
+      List<PinAttachment> attachments =
+          ref.read(selectedPinAttachmentsProvider);
+      var attachmentNotifier =
+          ref.read(selectedPinAttachmentsProvider.notifier);
+      if (attachments.isNotEmpty) {
+        attachments.addAll(newAttachments);
+        attachmentNotifier.update((state) => [...attachments]);
+      } else {
+        attachmentNotifier.update((state) => newAttachments);
+      }
+    }
+  }
+
+  // attachment selection sheet
+  static void showAttachmentSelection(
```

Comment:
yeah, pretty sure these should be on some kind of widget and the different onTap should be their own functions on it, making it a lot easier to read.